File: 380/main.py
from random import randint


class RandomizedSet:
    """
    approach:
    have a list to get random with O(1) time
    """

    def __init__(self):
        self.set = set()
        self.val_to_i = {}
        self.list = []

    def insert(self, val: int) -> bool:
        was_in_set = val in self.set
        self.set.add(val)
        if not was_in_set:
            self.val_to_i[val] = len(self.list)
            self.list.append(val)
        return not was_in_set

    def remove(self, val: int) -> bool:
        was_in_set = val in self.set
        if was_in_set:
            self.set.remove(val)
            i = self.val_to_i[val]
            self.list[i], self.list[len(self.list) - 1] = (
                self.list[len(self.list) - 1],
                self.list[i],
            )
            self.val_to_i[self.list[i]] = i
            del self.val_to_i[val]
            self.list.pop()
        return was_in_set

    def getRandom(self) -> int:
        return self.list[randint(0, len(self.list) - 1)]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

thing = RandomizedSet()
print(thing.insert(0))
print(thing.insert(1))
print(thing.remove(0))
print(thing.insert(2))
print(thing.remove(1))
print(thing.getRandom())

File: 380/main.ts
class RandomizedSet {
  mapOfIndices: Map<number, number>;
  values: number[];
  constructor() {
    this.mapOfIndices = new Map();
    this.values = [];
  }

  insert(val: number): boolean {
    let notInSet = !this.mapOfIndices.has(val);
    if (notInSet) {
      this.mapOfIndices.set(val, this.values.length);
      this.values.push(val);
    }
    // console.log(`add ${val}`);
    // console.log(this.mapOfIndices);
    // console.log(this.values);
    return notInSet;
  }

  remove(val: number): boolean {
    let inSet = this.mapOfIndices.has(val);
    if (inSet) {
      let index = this.mapOfIndices.get(val);
      let lastItem = this.values.pop();
      if (lastItem !== val) this.values[index] = lastItem;
      this.mapOfIndices.set(lastItem, index);
      this.mapOfIndices.delete(val);
    }
    // console.log(`remove ${val}`);
    // console.log(this.mapOfIndices);
    // console.log(this.values);
    return inSet;
  }

  getRandom(): number {
    return this.values[Math.floor(Math.random() * this.values.length)];
  }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */

// approach insert // basically pretty regular
Math.random();

// have a map and a list
// when inserting, insert into map (val, index)
// when removing,
// look for index of val
// remove from list
// swap with last val in list
// update map of last val's index to new index
